src/develop/blend.c

```diff
@@ -514,6 +577,175 @@ void dt_develop_blend_process(struct dt_iop_module_t *self, struct dt_dev_pixelp
+    dt_opencl_set_kernel_arg(devid, kernel_mask_blur, 1, sizeof(cl_mem), &out);
+    dt_opencl_set_kernel_arg(devid, kernel_mask_blur, 2, sizeof(int), &iwidth);
+    dt_opencl_set_kernel_arg(devid, kernel_mask_blur, 3, sizeof(int), &iheight);
+    dt_opencl_set_kernel_arg(devid, kernel_mask_blur, 4, sizeof(int), &c42);
+    dt_opencl_set_kernel_arg(devid, kernel_mask_blur, 5, sizeof(int), &c41);
+    dt_opencl_set_kernel_arg(devid, kernel_mask_blur, 6, sizeof(int), &c40);
+    dt_opencl_set_kernel_arg(devid, kernel_mask_blur, 7, sizeof(int), &c33);
+    dt_opencl_set_kernel_arg(devid, kernel_mask_blur, 8, sizeof(int), &c32);
+    dt_opencl_set_kernel_arg(devid, kernel_mask_blur, 9, sizeof(int), &c31);
+    dt_opencl_set_kernel_arg(devid, kernel_mask_blur, 10, sizeof(int), &c30);
+    dt_opencl_set_kernel_arg(devid, kernel_mask_blur, 11, sizeof(int), &c22);
+    dt_opencl_set_kernel_arg(devid, kernel_mask_blur, 12, sizeof(int), &c21);
+    dt_opencl_set_kernel_arg(devid, kernel_mask_blur, 13, sizeof(int), &c20);
+    dt_opencl_set_kernel_arg(devid, kernel_mask_blur, 14, sizeof(int), &c11);
+    dt_opencl_set_kernel_arg(devid, kernel_mask_blur, 15, sizeof(int), &c10);
+    dt_opencl_set_kernel_arg(devid, kernel_mask_blur, 16, sizeof(int), &c00);
+    const int err = dt_opencl_enqueue_kernel_2d(devid, kernel_mask_blur, sizes);
+    if(err != CL_SUCCESS) goto error;
+  }
+
+  {
+    size_t sizes[3] = { ROUNDUPWD(iwidth), ROUNDUPHT(iheight), 1 };
+    dt_opencl_set_kernel_arg(devid, kernel_write_mask, 0, sizeof(cl_mem), &out);
+    dt_opencl_set_kernel_arg(devid, kernel_write_mask, 1, sizeof(cl_mem), &tmp);
+    dt_opencl_set_kernel_arg(devid, kernel_write_mask, 2, sizeof(int), &iwidth);
+    dt_opencl_set_kernel_arg(devid, kernel_write_mask, 3, sizeof(int), &iheight);
+    const int err = dt_opencl_enqueue_kernel_2d(devid, kernel_write_mask, sizes);
+    if(err != CL_SUCCESS) goto error;
+  }  
+
+  {
+    const int err = dt_opencl_read_host_from_device(devid, lum, tmp, iwidth, iheight, sizeof(float));
+    if(err != CL_SUCCESS) goto error;
+  }
+
+  dt_opencl_free_kernel(kernel_calc_blend);
+  dt_opencl_free_kernel(kernel_write_mask);
+  dt_opencl_free_kernel(kernel_read_mask);
+  dt_opencl_free_kernel(kernel_mask_blur);
+
+  dt_opencl_release_mem_object(tmp);
+  dt_opencl_release_mem_object(blur);
+  dt_opencl_release_mem_object(out);
+  tmp = blur = out = NULL;
+
+  // here we have the slightly blurred full detail available
+  float *warp_mask = dt_dev_distort_luminance_mask(p, lum, self);
+  if(warp_mask == NULL) goto error;
+
+  const int msize = owidth * oheight;
+#ifdef _OPENMP
+  #pragma omp parallel for simd default(none) \
+  dt_omp_firstprivate(mask, warp_mask, msize) \
+  schedule(simd:static) aligned(mask, warp_mask : 64) 
+ #endif
+  for(int idx = 0; idx < msize; idx++)
+  {
+    mask[idx] = mask[idx] * warp_mask[idx];
+  }
+  dt_free_align(warp_mask);
+  return;
```

Comment:
here we are missing the free for lum. again maybe better to release memory after line 725 and remove the free on the `error:` section below.

Reply:
yeah, the free for lum was missing (thanks), in the error section it is required as we might fail in cl code.